sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -751,6 +753,43 @@ abstract class TypeCoercionBase {
     }
   }
 
+  /**
+   * Casts types according to bucket info for Equality expression.
+   */
+  object EqualityTypeCasts extends TypeCoercionRule {
+
+    private def isIntegralTypes(exprs: Seq[Expression]): Boolean = {
+      exprs.map(_.dataType).forall {
+        case _: IntegralType => true
+        case DecimalType.Fixed(_, 0) => true
+        case _ => false
+      }
+    }
+
+    override val transform: PartialFunction[Expression, Expression] = {
+      case b @ Equality(left: Attribute, right: Attribute)
+          if b.childrenResolved && left.dataType != right.dataType && isIntegralTypes(b.children) =>
```

Comment:
Why we don't care what `left.dataType` is?  The test added in `TypeCoercionSuite.scala` only tests Integer, Long and Decimal. If we don't have a constraint on the `left.dataType` probably we need to test more types?

Reply:
Oh Sorry I mis-read the code here: it check `isIntegralTypes(b.children)` in which `b` is the Equality thus both left and right data types are verified.